File: src/main/java/com/leijendary/spring/boot/template/core/config/DatabaseConfiguration.java
package com.leijendary.spring.boot.template.core.config;

import static com.leijendary.spring.boot.template.core.config.DatabaseConfiguration.DataSourceType.PRIMARY;
import static com.leijendary.spring.boot.template.core.config.DatabaseConfiguration.DataSourceType.READONLY;
import static org.springframework.transaction.support.TransactionSynchronizationManager.isCurrentTransactionReadOnly;

import java.util.HashMap;

import javax.sql.DataSource;

import com.leijendary.spring.boot.template.core.config.properties.DataSourcePrimaryProperties;
import com.leijendary.spring.boot.template.core.config.properties.DataSourceReadonlyProperties;
import com.zaxxer.hikari.HikariDataSource;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;
import org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy;
import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource;

import lombok.RequiredArgsConstructor;

@Configuration
@RequiredArgsConstructor
public class DatabaseConfiguration {

    private final DataSourcePrimaryProperties primaryProperties;
    private final DataSourceReadonlyProperties readonlyProperties;

    public enum DataSourceType {
        PRIMARY,
        READONLY;
    }

    public class TransactionRoutingDataSource extends AbstractRoutingDataSource {

        @Override
        protected Object determineCurrentLookupKey() {
            final var isReadOnly = isCurrentTransactionReadOnly();

            return isReadOnly ? READONLY : PRIMARY;
        }
    }

    @Bean
    @Primary
    public DataSource dataSource() {
        return new LazyConnectionDataSourceProxy(routingDataSource());
    }

    @Bean
    public TransactionRoutingDataSource routingDataSource() {
        final var routing = new TransactionRoutingDataSource();
        final var dataSource = new HashMap<>();
        dataSource.put(PRIMARY, primaryDataSource());
        dataSource.put(READONLY, readonlyDataSource());

        routing.setTargetDataSources(dataSource);

        return routing;
    }

    @Bean
    public HikariDataSource primaryDataSource() {
        return new HikariDataSource(primaryProperties);
    }

    @Bean
    public HikariDataSource readonlyDataSource() {
        return new HikariDataSource(readonlyProperties);
    }
}

File: src/main/kotlin/com/leijendary/spring/boot/template/core/config/DatabaseConfiguration.kt
package com.leijendary.spring.boot.template.core.config

import com.leijendary.spring.boot.template.core.config.DatabaseConfiguration.DataSourceType.PRIMARY
import com.leijendary.spring.boot.template.core.config.DatabaseConfiguration.DataSourceType.READONLY
import com.leijendary.spring.boot.template.core.config.properties.DataSourcePrimaryProperties
import com.leijendary.spring.boot.template.core.config.properties.DataSourceReadonlyProperties
import com.zaxxer.hikari.HikariDataSource
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy
import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource
import org.springframework.transaction.support.TransactionSynchronizationManager.isCurrentTransactionReadOnly
import javax.sql.DataSource

@Configuration
class DatabaseConfiguration(
    private val primaryProperties: DataSourcePrimaryProperties,
    private val readonlyProperties: DataSourceReadonlyProperties
) {
    enum class DataSourceType {
        PRIMARY,
        READONLY
    }

    inner class TransactionRoutingDataSource : AbstractRoutingDataSource() {
        override fun determineCurrentLookupKey(): Any {
            val isReadOnly = isCurrentTransactionReadOnly()

            return if (isReadOnly) READONLY else PRIMARY
        }
    }

    @Bean
    @Primary
    fun dataSource(): DataSource {
        return LazyConnectionDataSourceProxy(routingDataSource())
    }

    @Bean
    fun routingDataSource(): TransactionRoutingDataSource {
        val routing = TransactionRoutingDataSource()
        val dataSource = HashMap<Any, Any>()
        dataSource[PRIMARY] = primaryDataSource()
        dataSource[READONLY] = readonlyDataSource()

        routing.setTargetDataSources(dataSource)

        return routing
    }

    @Bean
    fun primaryDataSource(): HikariDataSource {
        return HikariDataSource(primaryProperties)
    }

    @Bean
    fun readonlyDataSource(): HikariDataSource {
        return HikariDataSource(readonlyProperties)
    }
}